connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java

```diff
@@ -1327,30 +1334,39 @@ public WorkerTask doBuild(Task task,
                     connectorClientConfigOverridePolicy, kafkaClusterId);
             KafkaProducer<byte[], byte[]> producer = new KafkaProducer<>(producerProps);
 
-            TopicAdmin topicAdmin;
+            // Prepare to create a topic admin if the task requires one, but do not actually create an instance
+            // until/unless one is needed
+            final AtomicReference<TopicAdmin> topicAdmin = new AtomicReference<>();
+            final Supplier<TopicAdmin> topicAdminCreator = () -> topicAdmin.updateAndGet(existingAdmin -> {
+                if (existingAdmin != null) {
+                    return existingAdmin;
+                }
+                Map<String, Object> adminOverrides = adminConfigs(id.connector(), "connector-adminclient-" + id, config,
+                        sourceConfig, connectorClass, connectorClientConfigOverridePolicy, kafkaClusterId, ConnectorType.SOURCE);
+                Admin adminClient = Admin.create(adminOverrides);
+                return new TopicAdmin(adminOverrides.get(BOOTSTRAP_SERVERS_CONFIG), adminClient);
+            });
+
             Map<String, TopicCreationGroup> topicCreationGroups;
             if (config.topicCreationEnable() && sourceConfig.usesTopicCreation()) {
                 topicCreationGroups = TopicCreationGroup.configuredGroups(sourceConfig);
                 // Create a topic admin that the task can use for topic creation
-                Map<String, Object> adminOverrides = adminConfigs(id.connector(), "connector-adminclient-" + id, config,
-                        sourceConfig, connectorClass, connectorClientConfigOverridePolicy, kafkaClusterId, ConnectorType.SOURCE);
-                topicAdmin = new TopicAdmin(adminOverrides);
+                topicAdminCreator.get();
```

Comment:
This call won't result in a leak because the `AbstractWorkerSourceTask` class will always close its `TopicAdmin` during shutdown; see [AbstractWorkerSourceTask::close](https://github.com/C0urante/kafka/blob/fe675683b4671714d02c8e674bd36347d0c345e9/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractWorkerSourceTask.java#L306-L308). And, given the idempotent nature of `Admin::close`, it's fine to close it from both the `AbstractWorkerSourceTask` and `ConnectorOffsetBackingStore` classes.

The underlying point about ownership being unclear is fair, though, and the control flow that creates the topic admin in this method is especially convoluted. I've added a couple comments that hopefully clear things up. I wish the logic itself could be better but as it is, things are easy to test and that's taken advantage of by adding a ton of coverage in the [WorkerTest suite](https://github.com/C0urante/kafka/blob/fe675683b4671714d02c8e674bd36347d0c345e9/connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerTest.java#L1592-L1653). Happy to refactor if there's an easily-testable approach that's clearer, though.

Reply:
Looking at the tests it doesn't seem we have any that assert that the `Admin` actually gets closed. `WorkerSourceTaskTest` asserts this, but for this code we're reliant that every invocation of `doBuild` always returns a `WorkerSourceTask` which is always stopped. I can see this is pretty hard to test though.

Reply:
👍 I added the logic for checking those conditions to `doBuild`. It's still a little clunky since we're checking `regularSourceTaskUsesConnectorSpecificOffsetsStore` in two places now (`doBuild` and `offsetStoreForRegularSourceTask`), but for testing purposes it's easier to work with since we can continue to unit test `offsetStoreForRegularSourceTask` independently without having to go through `doBuild`. And it's certainly more readable than the caching `Supplier` logic that it replaced.

Good point about ensuring that we stop `WorkerSourceTask` instances, we definitely do a poor job of that when we fail to build a task. In fact, this has already been documented as a bug with [an accompanying fix PR](https://github.com/apache/kafka/pull/11608). If that's worth pursuing at some point let me know and I can rebase onto the latest trunk.

Reply:
Thanks! When you have time it would be great if you could rebase that PR.

Reply:
👍 done.